### PR TITLE
[FLINK-2897] [runtime] Use distinct initial indices for OutputEmitter round-robin

### PR DESCRIPTION
**Pull request by greghogan**

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputEmitter.java [hunk 1]
```diff
@@ -143,14 +148,25 @@ public OutputEmitter(ShipStrategyType strategy, TypeComparator<T> comparator, Pa
 	
 	// --------------------------------------------------------------------------------------------
 
+	private int[] forward() {
+		if (this.channels == null) {
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputEmitter.java]: This method preserves the behavior of `FORWARD` only if `numberOfChannels` is always 0 when `FORWARD` is configured which is out of control of the `OutputEmitter`.
Should we add a check somewhere (e.g., in this condition to reduce the overhead) to ensure that `numberOfChannels` is acutally 0 for `FORWARD`?

Comment (greghogan) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputEmitter.java]: Yes, I did overlook that, though customPartition does likewise and the OutputEmitter will only be used for a single strategy. The strategy is set in the constructor so the channel list can be created there for the types which do not broadcast.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputEmitter.java [hunk 2]
```diff
@@ -143,16 +148,24 @@ public OutputEmitter(ShipStrategyType strategy, TypeComparator<T> comparator, Pa
...
+	}
+
 	private int[] robin(int numberOfChannels) {
-		if (this.channels == null || this.channels.length != 1) {
-			this.channels = new int[1];
+		int nextChannel = this.nextChannelToSendTo;
+
+		if (nextChannel >= numberOfChannels) {
+			if (nextChannel == numberOfChannels) {
+				nextChannel = 0;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputEmitter.java]: Shouldn't this case be also covered by `nextChannel %= numberOfChannels`?

**Comment by StephanEwen** [comment]
Aside from till's comment (simplification of the "robin()" case), this looks good.

**Comment by tillrohrmann** [comment]
Yes, +1 for merging. Thanks for your contribution @greghogan.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputEmitter.java (hunk 2 above)
Comment (greghogan) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputEmitter.java]: The modulo computation is executed in two cases: 1) the initial call to robin when the source parallelism is greater than the receiver parallelism and 2) when the receiver parallelism is dynamically changing.

Unoptimized modulo is much slower than a test for equality but is required in the first case to evenly distribute the first outputs.

**Comment by StephanEwen** [comment]
Will merge this...

**Pull request closed without merging**